// baev3-log/gen-nodejs/logchain.js
//
// Autogenerated by Thrift Compiler (0.9.1)
//
// DO NOT EDIT UNLESS YOU ARE SURE THAT YOU KNOW WHAT YOU ARE DOING
//
var Thrift = require('thrift').Thrift;

var ttypes = require('./user_types');
//HELPER FUNCTIONS AND STRUCTURES

logchain_log_args = function(args) {
  this.log = null;
  if (args) {
    if (args.log !== undefined) {
      this.log = args.log;
    }
  }
};
logchain_log_args.prototype = {};
logchain_log_args.prototype.read = function(input) {
  input.readStructBegin();
  while (true)
  {
    var ret = input.readFieldBegin();
    var fname = ret.fname;
    var ftype = ret.ftype;
    var fid = ret.fid;
    if (ftype == Thrift.Type.STOP) {
      break;
    }
    switch (fid)
    {
      case 1:
      if (ftype == Thrift.Type.STRUCT) {
        this.log = new ttypes.BaeLog();
        this.log.read(input);
      } else {
        input.skip(ftype);
      }
      break;
      case 0:
        input.skip(ftype);
        break;
      default:
        input.skip(ftype);
    }
    input.readFieldEnd();
  }
  input.readStructEnd();
  return;
};

logchain_log_args.prototype.write = function(output) {
  output.writeStructBegin('logchain_log_args');
  if (this.log !== null && this.log !== undefined) {
    output.writeFieldBegin('log', Thrift.Type.STRUCT, 1);
    this.log.write(output);
    output.writeFieldEnd();
  }
  output.writeFieldStop();
  output.writeStructEnd();
  return;
};

logchain_log_result = function(args) {
  this.success = null;
  if (args) {
    if (args.success !== undefined) {
      this.success = args.success;
    }
  }
};
logchain_log_result.prototype = {};
logchain_log_result.prototype.read = function(input) {
  input.readStructBegin();
  while (true)
  {
    var ret = input.readFieldBegin();
    var fname = ret.fname;
    var ftype = ret.ftype;
    var fid = ret.fid;
    if (ftype == Thrift.Type.STOP) {
      break;
    }
    switch (fid)
    {
      case 0:
      if (ftype == Thrift.Type.I32) {
        this.success = input.readI32();
      } else {
        input.skip(ftype);
      }
      break;
      case 0:
        input.skip(ftype);
        break;
      default:
        input.skip(ftype);
    }
    input.readFieldEnd();
  }
  input.readStructEnd();
  return;
};

logchain_log_result.prototype.write = function(output) {
  output.writeStructBegin('logchain_log_result');
  if (this.success !== null && this.success !== undefined) {
    output.writeFieldBegin('success', Thrift.Type.I32, 0);
    output.writeI32(this.success);
    output.writeFieldEnd();
  }
  output.writeFieldStop();
  output.writeStructEnd();
  return;
};

logchain_check_version_args = function(args) {
  this.version = null;
  if (args) {
    if (args.version !== undefined) {
      this.version = args.version;
    }
  }
};
logchain_check_version_args.prototype = {};
logchain_check_version_args.prototype.read = function(input) {
  input.readStructBegin();
  while (true)
  {
    var ret = input.readFieldBegin();
    var fname = ret.fname;
    var ftype = ret.ftype;
    var fid = ret.fid;
    if (ftype == Thrift.Type.STOP) {
      break;
    }
    switch (fid)
    {
      case -1:
      if (ftype == Thrift.Type.STRING) {
        this.version = input.readString();
      } else {
        input.skip(ftype);
      }
      break;
      case 0:
        input.skip(ftype);
        break;
      default:
        input.skip(ftype);
    }
    input.readFieldEnd();
  }
  input.readStructEnd();
  return;
};

logchain_check_version_args.prototype.write = function(output) {
  output.writeStructBegin('logchain_check_version_args');
  if (this.version !== null && this.version !== undefined) {
    output.writeFieldBegin('version', Thrift.Type.STRING, -1);
    output.writeString(this.version);
    output.writeFieldEnd();
  }
  output.writeFieldStop();
  output.writeStructEnd();
  return;
};

logchain_check_version_result = function(args) {
  this.success = null;
  if (args) {
    if (args.success !== undefined) {
      this.success = args.success;
    }
  }
};
logchain_check_version_result.prototype = {};
logchain_check_version_result.prototype.read = function(input) {
  input.readStructBegin();
  while (true)
  {
    var ret = input.readFieldBegin();
    var fname = ret.fname;
    var ftype = ret.ftype;
    var fid = ret.fid;
    if (ftype == Thrift.Type.STOP) {
      break;
    }
    switch (fid)
    {
      case 0:
      if (ftype == Thrift.Type.I32) {
        this.success = input.readI32();
      } else {
        input.skip(ftype);
      }
      break;
      case 0:
        input.skip(ftype);
        break;
      default:
        input.skip(ftype);
    }
    input.readFieldEnd();
  }
  input.readStructEnd();
  return;
};

logchain_check_version_result.prototype.write = function(output) {
  output.writeStructBegin('logchain_check_version_result');
  if (this.success !== null && this.success !== undefined) {
    output.writeFieldBegin('success', Thrift.Type.I32, 0);
    output.writeI32(this.success);
    output.writeFieldEnd();
  }
  output.writeFieldStop();
  output.writeStructEnd();
  return;
};

logchainClient = exports.Client = function(output, pClass) {
    this.output = output;
    this.pClass = pClass;
    this.seqid = 0;
    this._reqs = {};
};
logchainClient.prototype = {};
logchainClient.prototype.log = function(log, callback) {
	console.log("logchain.log");
  this.seqid += 1;
  this._reqs[this.seqid] = callback;
  this.send_log(log);
};

logchainClient.prototype.send_log = function(log) {
	console.log("logchain send_log = " + log);
  var output = new this.pClass(this.output);
  output.writeMessageBegin('log', Thrift.MessageType.CALL, this.seqid);
  var args = new logchain_log_args();
  args.log = log;
  args.write(output);
  output.writeMessageEnd();
  return this.output.flush();
};

logchainClient.prototype.recv_log = function(input,mtype,rseqid) {
	console.log("logchain recv_log");
  var callback = this._reqs[rseqid] || function() {};
  delete this._reqs[rseqid];
  if (mtype == Thrift.MessageType.EXCEPTION) {
    var x = new Thrift.TApplicationException();
    x.read(input);
    input.readMessageEnd();
    return callback(x);
  }
  var result = new logchain_log_result();
  result.read(input);
  input.readMessageEnd();
console.log("result.success = " + result.success);
  if (null !== result.success) {
    return callback(null, result.success);
  }
  return callback('log failed: unknown result');
};
logchainClient.prototype.check_version = function(version, callback) {
  this.seqid += 1;
  this._reqs[this.seqid] = callback;
  this.send_check_version(version);
};

logchainClient.prototype.send_check_version = function(version) {
  var output = new this.pClass(this.output);
  output.writeMessageBegin('check_version', Thrift.MessageType.CALL, this.seqid);
  var args = new logchain_check_version_args();
  args.version = version;
  args.write(output);
  output.writeMessageEnd();
  return this.output.flush();
};

logchainClient.prototype.recv_check_version = function(input,mtype,rseqid) {
  var callback = this._reqs[rseqid] || function() {};
  delete this._reqs[rseqid];
  if (mtype == Thrift.MessageType.EXCEPTION) {
    var x = new Thrift.TApplicationException();
    x.read(input);
    input.readMessageEnd();
    return callback(x);
  }
  var result = new logchain_check_version_result();
  result.read(input);
  input.readMessageEnd();

  if (null !== result.success) {
    return callback(null, result.success);
  }
  return callback('check_version failed: unknown result');
};
logchainProcessor = exports.Processor = function(handler) {
  this._handler = handler
}
logchainProcessor.prototype.process = function(input, output) {
  var r = input.readMessageBegin();
  if (this['process_' + r.fname]) {
    return this['process_' + r.fname].call(this, r.rseqid, input, output);
  } else {
    input.skip(Thrift.Type.STRUCT);
    input.readMessageEnd();
    var x = new Thrift.TApplicationException(Thrift.TApplicationExceptionType.UNKNOWN_METHOD, 'Unknown function ' + r.fname);
    output.writeMessageBegin(r.fname, Thrift.MessageType.Exception, r.rseqid);
    x.write(output);
    output.writeMessageEnd();
    output.flush();
  }
}

logchainProcessor.prototype.process_log = function(seqid, input, output) {
  var args = new logchain_log_args();
  args.read(input);
  input.readMessageEnd();
  this._handler.log(args.log, function (err, result) {
    var result = new logchain_log_result((err != null ? err : {success: result}));
    output.writeMessageBegin("log", Thrift.MessageType.REPLY, seqid);
    result.write(output);
    output.writeMessageEnd();
    output.flush();
  })
}

logchainProcessor.prototype.process_check_version = function(seqid, input, output) {
  var args = new logchain_check_version_args();
  args.read(input);
  input.readMessageEnd();
  this._handler.check_version(args.version, function (err, result) {
    var result = new logchain_check_version_result((err != null ? err : {success: result}));
    output.writeMessageBegin("check_version", Thrift.MessageType.REPLY, seqid);
    result.write(output);
    output.writeMessageEnd();
    output.flush();
  })
}
